consensus/src/liveness/unequivocal_proposer_election.rs

```diff
@@ -0,0 +1,89 @@
+// Copyright (c) Aptos
+// SPDX-License-Identifier: Apache-2.0
+
+use std::cmp::Ordering;
+
+use aptos_crypto::HashValue;
+use aptos_infallible::Mutex;
+use aptos_logger::{error, SecurityEvent};
+use consensus_types::{
+    block::Block,
+    common::{Author, Round},
+};
+
+use super::proposer_election::ProposerElection;
+
+// Wrapper around ProposerElection.
+//
+// Provides is_valid_proposal that remembers, and rejects if
+// the same leader proposes multiple blocks.
+pub struct UnequivocalProposerElection {
+    proposer_election: Box<dyn ProposerElection + Send + Sync>,
+    already_proposed: Mutex<(Round, HashValue)>,
+}
+
+impl ProposerElection for UnequivocalProposerElection {
+    fn get_valid_proposer(&self, round: Round) -> Author {
+        self.proposer_election.get_valid_proposer(round)
+    }
+}
+
+impl UnequivocalProposerElection {
+    pub fn new(proposer_election: Box<dyn ProposerElection + Send + Sync>) -> Self {
+        Self {
+            proposer_election,
+            already_proposed: Mutex::new((0, HashValue::zero())),
+        }
+    }
+
+    // Return if a given proposed block is valid:
+    // - if a given author is a valid candidate for being a proposer
+    // - if this is the first block proposer has submitted in this round
+    // - if it is not old proposal
+    pub fn is_valid_proposal(&self, block: &Block) -> bool {
+        block.author().map_or(false, |author| {
+            let valid_author = self.is_valid_proposer(author, block.round());
+            if !valid_author {
+                error!(
+                    SecurityEvent::InvalidConsensusProposal,
```

Comment:
this is not a security event though, it's possible that just this node has a different committed state so it derives another leader for the round